Distribute Sampler data in parallel

aeslaughter opened the pull request:
(closes #13906)

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
New interface for Sampler objects to allow for global and/or local retrival

(closes #13906)

aeslaughter committed:
Create test for memory use of Sampler get methods

aeslaughter committed:
Update stochstic tools objects to use getLocalSamples

aeslaughter committed:
Improve API and documentation for Sampler

aeslaughter committed:
Update documentation

aeslaughter committed:
Update SQA requirements for stochastic tools

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit ee06a775fc6ef6ed7096392e355e18e0dadb23c2
+Subproject commit 12809bf734da70025424972d3c294e7604938336
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/413289/) on bcf55d0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14191/site)

This comment will be updated on new commits.

aeslaughter committed:
Update tests for SQA and python package restriction

jiangwen84 commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,149 @@
+   * getSamples() or getLocalSamples() methods.
    */
-  dof_id_type getLocalRowEnd();
+  virtual void sampleSetUp(){};
```
Could you add comments for these two functions?

aeslaughter replied:
The comments exist, the diff just doesn't show them.

jiangwen84 commented on framework/include/distributions/DistributionInterface.h:
```diff
@@ -37,6 +37,12 @@ class DistributionInterface
+  /**
+   * Return the list of distribution names.
+   */
+  // const std::vector<DistributionName> & getDistributionNames() const { return
```
delete commented out codes

jiangwen84 commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,149 @@
- * The main method in this object is the getSamples method which returns the distribution samples.
- * This method will return the same results for each call when the "_generator" from this class
- * is used by child classes.
+ * The main methods in this object is the getSamples/getLocalSamples methods which returns the
```
returns -> return

jiangwen84 commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,149 @@
+   * @return A double for the random number, this is double because MooseRandom class uses double.
    */
-  dof_id_type getLocalNumerOfRows();
+  double rand(unsigned int index = 0);
```
could you change the function name to differentiate c++'s own rand()?

aeslaughter replied:
Yes, what do you suggest?

aeslaughter replied:
How about `getRand`?

jiangwen84 replied:
great

jiangwen84 commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,149 @@
+   * set methods can be called within the constructors of child objects, see
+   * FEProblemBase::addSampler method.
+   *
+   * This method is for internal use by MOOSE, it should not be called otherwise.
```
similar to line 144, can be deleted?

jiangwen84 commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,149 @@
+   * Store the state of the MooseRandom generator so that new calls to getSamples/getLocalSamples
+   * methods will create new numbers. This also handles sample data cache, if enabled.
+   *
+   * This method is for internal use by MOOSE, it should not be called otherwise.
```
similar to line 157-158, can be deleted?

jiangwen84 commented on modules/stochastic_tools/include/transfers/StochasticToolsTransfer.h:
```diff
@@ -31,4 +31,9 @@ class StochasticToolsTransfer : public MultiAppTransfer
 
   virtual void finalizeToMultiapp();
+
+  void setGlobalMultiAppIndex(dof_id_type index) { _global_index = index; }
```
add comments?

jiangwen84 commented on modules/stochastic_tools/src/transfers/SamplerTransfer.C:
```diff
@@ -94,23 +94,27 @@ SamplerTransfer::execute()
 void
 SamplerTransfer::finalizeToMultiapp()
 {
+  _samples.resize(0, 0);
```
why to resize?

aeslaughter replied:
frees up the memory, I don't want to keep copies of the sampled data around. This is only in batch mode, where the sample data is generated in initialize, utilized in execute, and then cleaned up in finalize

jiangwen84 commented on modules/stochastic_tools/test/tests/multiapps/batch_commandline_control/gold/master_multiple_out_storage_0002.csv:
```diff
@@ -1,4 +1,4 @@
-6.82795
-5.96624
+sample
+8.9708
```
do you know why all random numbers are changed?

aeslaughter replied:
Yes, as you discovered the current implementation calls getSamples once during initialization, then save state, which is wrong. This fixes that problem.

jiangwen84 replied:
👍

jiangwen84 reviewed (changes requested):
looks great, just some minor comments.

aeslaughter commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3249,7 +3250,10 @@ FEProblemBase::execute(const ExecFlagType & exec_type)
+  // construction make the first execute() call. This ensures that the random number generator
+  // object is the correct state prior to any other object (e.g., Transfers) attempts to extract
+  // data from the Sampler.
   if (exec_type != EXEC_INITIAL)
```
No, EXEC_INITIAL is still called, it is just called elsewhere. I will improve the docs.

aeslaughter committed:
Address review comments

aeslaughter committed:
Change Sampler::rand to getRand to avoid confusion with std::rand

aeslaughter commented:
@jiangwen84 Thanks for the comments, they should all be addressed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/415171/) on 0798358 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e4ad85f3076cd5c2d50aafb8d118c1a632e82865`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/415208/) on be98b8f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4d49c33ebe396414bf60650a882016eca1e1fd9c`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit ee06a775fc6ef6ed7096392e355e18e0dadb23c2
+Subproject commit a93d9ac707babeda32ac3a01f710718f18a73a85
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/415446/) on fa350f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4d49c33ebe396414bf60650a882016eca1e1fd9c`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/415446/) on fa350f2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4d49c33ebe396414bf60650a882016eca1e1fd9c`

aeslaughter commented:
This is still waiting for updates to Grizzly

aeslaughter committed:
Create row iterator methods for Sampler object

(refs #13906)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/419066/) on 0434454 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 92dcf2844cf1e40fed5813cddb68adb1e04d8790`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/419844/) on 9bf540e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9600adbdef07e974a6145298bcde263d2f5d87eb`

aeslaughter committed:
WIP: Add matrix/vector size limits to Sampler

(refs #13906)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/419946/) on a846827 : invalidated by @milljm

updated password on mooseframework

aeslaughter committed:
Update SamplerData for getNextLocalRow

(refs #13906)

aeslaughter committed:
Update SamplerTransfer to use getNextLocalRow

(refs #13906)

aeslaughter committed:
Update MultiAppCommandLineControl to use getNextLocalRow

(refs #13906)

aeslaughter committed:
Update SamplerFullSolveMultiApp to use getNextLocalRow

(refs #13906)

aeslaughter committed:
Test errors within Sampler base class

(refs #13906)

aeslaughter committed:
Improve interface for python util for Sampler memory data

(refs #13906)

aeslaughter committed:
Add row compute method to API and improve documentation

(refs #13906)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/421431/) on 31f9bd7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1869f27065a1b32b97cdb50301cde134d86eb4ef`

aeslaughter committed:
Add perf graph to Sampler and Distribution objects

(refs #13906)

aeslaughter committed:
Update stochastic tools documentation to include getNextLocalRow

(refs #13906)

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit a93d9ac707babeda32ac3a01f710718f18a73a85
+Subproject commit eadd0a3f70521ce5557d8c328cc9b102338d2177
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

aeslaughter commented:
@bwspenc @jiangwen84 This should be ready for another review. I expanded the API to include a few more virtual methods, including a function that computes a row of data. 

I also add PerfGraph calls and found that the "advanceGenerators" takes about 10% of the run time for large number of samples.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/422007/) on 39bccf3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14191/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14191/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1869f27065a1b32b97cdb50301cde134d86eb4ef`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/422045/) on 8207918 : invalidated by @permcody

Trying out new submodule logic

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/422063/) on a28ee73 : invalidated by @permcody

moosebuild commented:
Job [Boost enabled, Pthreads only](https://civet.inl.gov/job/422066/) on a28ee73 : invalidated by @aeslaughter

I can&#39;t reproduce this regularly with moosebuild, I saw it 1 time only on many runs.

aeslaughter committed:
Remove copy of test that ran with threads

(refs #13906)

The test removed isn't needed, it is executed with threads by the test system

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/421968/) on d00ce27 : invalidated by @permcody

Relap-7 updated

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/421968/) on d00ce27 : invalidated by @permcody

Relap-7 updated

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/422071/) on a28ee73 : invalidated by @permcody

Relap-7 updated

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/422209/) on 6bc9d01 : invalidated by @permcody

Relap-7 updated

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/421971/) on d00ce27 : invalidated by @permcody

Relap-7 updated

bwspenc commented on framework/include/samplers/Sampler.h:
```diff
@@ -16,172 +16,204 @@
+   * the matrix to be created.
    */
-  void execute();
+  void setNumberOfRows(dof_id_type n_rows);
```
@aeslaughter In my case, I depend on a UserObject to compute the number of samples, and I can't do the `getUserObject()` call in the sampler's constructor. If I had an `initialSetup()` method I could call `setNumberOfRows()` there.

bwspenc replied:
Could `Sampler` derive from `SetupInterface`?

aeslaughter replied:
As discussed, I will try to move init() to the first Sampler::execute() call.

aeslaughter replied:
@bwspenc That doesn't work. It needs to happen in the constructor to allow for the correct number of MultiApp objects to be initialized. We are going to need to think about this more.

aeslaughter committed:
Update Sampler API to use getGlobalSamples to avoid ambiguity

(refs #13906)

aeslaughter committed:
Add legacy API for short-term application support

(refs #13906)

aeslaughter commented:
I have a small change for BISON ready once this is merged.

aeslaughter committed:
Disable recover for python script testing

(refs #13906)

bwspenc commented on framework/include/samplers/Sampler.h:
```diff
@@ -18,168 +18,163 @@
+   * Method for advancing the random number generator(s) by the supplied number or calls to rand().
    *
-   * @return The list of samples for the Sampler.
+   * TODO: This should be updated if the If the random number generator is updated to type that
```
This still needs to be fixed sometime.

aeslaughter commented:
@bwspenc @jiangwen84 Thanks for your help on getting this merged and the initial effort. This whole system and interface has vastly improved.

@friedmud Thanks for the next row idea, that was the a huge improvement to this system.

permcody commented on framework/include/samplers/Sampler.h:
```diff
@@ -190,18 +228,45 @@ class Sampler : public MooseObject, public SetupInterface, public DistributionIn
   dof_id_type _local_row_end;
+
+  /// Total number of rows in the sample matrix
+  dof_id_type _n_rows;
```
This is generally not the right type for length type holders. Our desktop build still defaults to 32-bit while container types can hold 64-bit counters.

permcody commented on modules/stochastic_tools/include/samplers/SobolSampler.h:
```diff
@@ -24,17 +24,22 @@ class SobolSampler : public Sampler
+  const std::vector<DistributionName> & _distribution_names;
+
+  /// The number of rows per matrix
+  const dof_id_type _num_rows_per_matrix;
```
Not the right type

permcody commented on modules/stochastic_tools/src/controls/MultiAppCommandLineControl.C:
```diff
@@ -94,50 +94,47 @@ void
+  else
+  {
+    std::ostringstream oss;
+    for (dof_id_type col = 0; col < _sampler.getNumberOfCols(); ++col)
```
type

permcody commented on modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:
```diff
@@ -90,27 +94,31 @@ SamplerFullSolveMultiApp::solveStepBatch(Real dt, Real target_time, bool auto_ad
+    }
 
-    if (i != num_items - 1)
+    if (i < _sampler.getLocalRowEnd() - 1)
```
Are you sure about this? This would terminate 2 from the end maybe?

permcody commented on modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:
```diff
@@ -145,28 +153,17 @@ SamplerFullSolveMultiApp::getCommandLineArgsParamHelper(unsigned int local_app)
+      MooseUtils::split(FullSolveMultiApp::getCommandLineArgsParamHelper(local_app), ";");
 
-  for (unsigned int i = 0; i < samples.size(); ++i)
+  for (dof_id_type col = 0; col < _sampler.getNumberOfCols(); ++col)
```
type

permcody commented on modules/stochastic_tools/src/samplers/SobolSampler.C:
```diff
@@ -17,58 +18,62 @@ validParams<SobolSampler>()
+  _a_matrix.resize(_num_rows_per_matrix, getNumberOfCols());
+  _b_matrix.resize(_num_rows_per_matrix, getNumberOfCols());
+  for (dof_id_type i = 0; i < _num_rows_per_matrix; ++i)
+    for (dof_id_type j = 0; j < getNumberOfCols(); ++j)
```
types

permcody commented on modules/stochastic_tools/src/transfers/SamplerTransfer.C:
```diff
@@ -56,35 +56,25 @@ SamplerTransfer::SamplerTransfer(const InputParameters & parameters)
 
   // Loop over all sub-apps
-  for (unsigned int app_index = 0; app_index < _multi_app->numGlobalApps(); app_index++)
+  for (dof_id_type row_index = _sampler_ptr->getLocalRowBegin();
```
type - last time I'll mention it

permcody reviewed (commented):
I know this is already merged. Just thought I'd go through really quickly to look for type errors. One general comment is that you decided to use `dof_id_type` as a catch all type. I think if you are going to do this you probably should just stick with `std::size_t` since the size of that is fixed. It's possible, although unlikely that using `dof_id_type` in place of a real 64-bit type (i.e. all of the vector size stuff) could cause an issue in practice depending on your PETSc configuration.
